### ISSUE TITLE
WAL/retention configuration for prometheus doesn't match disk size configuration

### ISSUE THREAD

**Issue by voelzmo** [gardener/gardener#4872]
**Summary**
We've seen cases where the prometheus WAL file filled up huge parts of the persistent disk, up to a point where the StatefulSet became unhealthy, because the prometheus container went into CrashLoopBackoff. Solutions in the past have been to delete the WAL in order to free up disk space. We want to find a way to prevent manual interaction from being necessary here.

**What needs to be done**
Find a way to prevent the WAL filling up the entire disk. Possible solutions could be
* configure the WAL retention, such that we can keep the current disk size (maybe because the disk space necessary with the current configuration is too expensive)
* configure the disk size, such that we can keep the current retention (maybe because we promised users certain qualities which require the current retention setting)
* find some middle ground between additional cost and operational qualities

**Background**
Prometheus uses the write-ahead-log file to write checkpoints to disk, such that it can recover from a crash. The size of the WAL is limited by a retention policy in size or time. [According to the official documentation](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects), the necessary disk space can be computed with 
```
needed_disk_space = retention_time_seconds * ingested_samples_per_second * bytes_per_sample
```

**Comment by rfranzke** [comment]
/area monitoring
cc @istvanballok

**Comment by istvanballok** [comment]
> The size of the WAL is limited by a retention policy in size or time.

Unfortunately, as far as I understood, the retention policy is only applied to the "persistent blocks", I don't think there is a way to limit the size of the write ahead log. The write ahead log captures the ingested data in 2 hour blocks and its size depends on the ingested data.

We could consider to delete the WAL before starting up prometheus, if its size is too big / there isn't much free disk space, ...

**Comment by rfranzke** [comment]
Any suggestions how to proceed here?

**Comment by ialidzhikov** [comment]
> Potential fix for https://github.com/gardener/gardener/issues/4872. (The issue didn't specify which prometheus was affected.)

@istvanballok I assume that this issue is about the prometheus that is part of the Shoot control plane. Is it possible to introduce similar retention for it as done in https://github.com/gardener/gardener/pull/5450 for the aggregate prometheus?

I think recently @kris94 and @mandelsoft hit this issue (exhausted disk space for the prometheus PV):
```
prometheus level=error ts=2022-05-13T07:27:53.015Z caller=main.go:915 err="opening storage failed: open /var/prometheus/data/wal/00000881: no space left on device" 
```

```console
$ dh -h | grep pv-shoot--foo--bar-cc2c7cc1-ea6c-4473-b14a-fa19f7c4ec4b
/dev/sdd         20G   20G     0 100% /host/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pv-shoot--foo--bar-cc2c7cc1-ea6c-4473-b14a-fa19f7c4ec4b/globalmount               
```

**Comment by istvanballok** [comment]
The prometheus in the shoot control plane already has a size based retention policy configuration:
https://github.com/gardener/gardener/blob/bc1827317538914af9be133cba42dfe7bb806e24/charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml#L76-L77
and I checked the internal issue you mentioned, and the configuration of that specific prometheus instance seems to be correct / as expected.

I'll try to check why the disk was still full. One possible explanation would be the write ahead log, as I think the retention policy does not apply to that. See [comment](https://github.com/gardener/gardener/issues/4872#issuecomment-999393480) above.

**Comment by istvanballok** [comment]
After having looked into the details of the most recent occurrence above, I think the root cause could be a bug in Prometheus that has been resolved with v2.33.4. After upgrading Prometheus with #5984, we can close this issue.
